### ISSUE TITLE
[Core] Actors not cleaning up resources correct because `force_kill=true`.

### ISSUE THREAD

**Issue by cadedaniel** [ray-project/ray#34124]
Discovered in the investigation of https://github.com/ray-project/ray/issues/31451 and https://github.com/ray-project/ray/pull/33976.

TL;DR things like actor destructors or atexit handlers are not guaranteed to be executed when we destroy actors. This is because we use `force_kill=true` in [gcs_actor_manager]((https://github.com/ray-project/ray/blob/bb24c16b63578c7ba2cc2f250ef9aad2be2278cc/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L749-L752)).

Ideally, we should send SIGTERM to worker processes so that they clean up any important state. After some time period, if the process has not already died already, we will then send a SIGKILL.

More information in Sang's comment here https://github.com/ray-project/ray/pull/33976#issuecomment-1495949518

**Comment by clarng** [comment]
but if os oom killer kick in these handlers won't get called right? we cannot guarantee os oom killer will never kick in

**Comment by cadedaniel** [comment]
Does our OOM killer immediately use SIGKILL? Or is there a SIGTERM first?

**Comment by clarng** [comment]
sigkill
if we wait then it might trigger os oom (which will force kill anyway)

**Comment by clarng** [comment]
trying to clean resources up when the machine is under pressure may not be the best idea

what are we trying to clean up? can that be done async?

**Comment by rkooo567** [comment]
@clarng this is only for the general shutdown cases. I think OOM killer sending SIGKILL is reasonable, and this just means we cannot properly trigger shutdown handler (that's the point of SIKILL actually)

**Comment by rkooo567** [comment]
Your case will be handled by https://github.com/ray-project/ray/issues/34125

**Comment by rkooo567** [comment]
https://github.com/ray-project/ray/issues/32952 -> this issue has a minimal repro script we should test

**Comment by mlguruz** [comment]
I believe this is also related:

I kept running into this when using `xgboost_ray`:
```
Exception ignored in: <function PlacementGroupResourceManager.__del__ at 0x7f6722c8a550>
File "/home/ubuntu/zstar/lib/python3.8/site-packages/ray/air/execution/resources/placement_group.py", line 211, in __del__
File "/home/ubuntu/zstar/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py", line 466, in _resume_span
File "/home/ubuntu/zstar/lib/python3.8/site-packages/ray/air/execution/resources/placement_group.py", line 193, in clear
File "/home/ubuntu/zstar/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 98, in wrapper
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function PlacementGroupResourceManager.__del__ at 0x7efecd343550>
```

**Comment by rkooo567** [comment]
Hmm, I believe it is actually a different issue. I think when you execute `__del__`, any module call is not guaranteed to succeed because modules can be GC'ed before the class is destructed.

cc @krfricke it is something `PlacementGroupResourceManager` should fix in its layer.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
